### PR TITLE
mempool janitor: periodic sweep and clean of not-confirming transactions

### PR DESCRIPTION
**Pull request by jgarzik**
The mempool janitor ("poolman") is a thread that runs every -janitorinterval
seconds.  The janitor scans and removes memory pool transactions
older than current time minus -janitorexpire seconds.

By default, janitor runs every 24 hrs, expiring TXs older than 72 hrs [and have failed to make it into a block in that time].

IsMine() transactions are not touched.

This is _intentionally crude_: easily reviewed, reasoned and tested; fitting easily within the current framework, or being backported to an older bitcoind.  A key goal was _not_ rewriting mempool.

Comments:
- One alternative implementation was considered:

```
     while (mempool byte size > limit)
            expire oldest !ismine
```

This would work, but require a sort step, as mempool is not time-ordered.  There is also the question of how often to run such a while{} loop, likely leading to some sort of high-water/low-water system.

The sweep implementation presented seemed more straightforward.

### CODE REVIEW

**Comment by ABISprotocol** [comment]
I just want to leave this as a comment rather than forking, pull requesting or any such-ness atm. I wanted to suggest that when this goes through to final at such point when rebase please consider the original remark I had some while back which was, I think, to rename from poolman.h to poolbeing.h, (see line 99 at src/Makefile.am for example).
Similarly poolman.cpp would become poolbeing.cpp and so forth.
Thank you if possible.  Please e-mail me with any questions.

File: src/init.cpp [hunk 1]
```diff
@@ -253,6 +254,8 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += "  -proxy=<ip:port>       " + _("Connect through SOCKS5 proxy") + "\n";
     strUsage += "  -seednode=<ip>         " + _("Connect to a node to retrieve peer addresses, and disconnect") + "\n";
     strUsage += "  -timeout=<n>           " + _("Specify connection timeout in milliseconds (default: 5000)") + "\n";
+    strUsage += "  -janitorinterval=<n>   " + _("Number of seconds between each mempool janitor run (default: 1 day)") + "\n";
+    strUsage += "  -janitorexpire=<n>     " + _("Number of seconds transactions live in memory pool, before removal (default: 3 days)") + "\n";
```
Comment (Diapolo) [inline src/init.cpp]: Nit: Could you switch these two and place them above ``-listen`` please. Also @luke-jr A few weeks ago changed all strings in help messages to use ``strprintf(_())`` as that makes life for translators easier.

File: src/init.cpp [hunk 2]
```diff
@@ -724,6 +727,16 @@ bool AppInit2(boost::thread_group& threadGroup)
             threadGroup.create_thread(&ThreadScriptCheck);
     }
 
+    // mempool janitor execution interval.  default interval: 1 day
+    int janitorInterval = GetArg("-janitorinterval", (60 * 60 * 24 * 1));
```
Comment (Diapolo) [inline src/init.cpp]: This allows negative numbers, is this intended?

File: src/poolman.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
+
```
Comment (Diapolo) [inline src/poolman.cpp]: Missing license :).

File: src/poolman.h [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#ifndef __BITCOIN_POOLMAN_H__
```
Comment (Diapolo) [inline src/poolman.h]: Also misses license and correct header define ``#ifndef BITCOIN_POOLMAN_H`` (and at the file end).

File: src/rpcblockchain.cpp [hunk 5]
```diff
@@ -406,6 +407,24 @@ Value gettxout(const Array& params, bool fHelp)
...
 }
 
+Value sweepmempool(const Array& params, bool fHelp)
+{
+    if (fHelp || params.size() != 0)
+        throw runtime_error(
+            "sweepmempool\n"
+            "\nRemoves old, unconfirmed transactions from mempool\n"
+            "\nResult:\n"
+            "n    (bool) true\n"
```
Comment (Diapolo) [inline src/rpcblockchain.cpp]: Nit: Is this correct help text formating?

**Comment by NanoAkron** [comment]
What is a Poolbeing? Politically-correct nonsense which doesn't belong in bitcoin core is what.

I vote Poolbot or Poolcleaner if you're so upset by the ending '-man' in the English language.

**Comment by ABISprotocol** [comment]
Hi, my request for poolbeing was alluding for something beyond the standard anthropomorphic reference, and yes, that which ended in man.  Not that I was upset by it, because I'm not, but I think that in light of the progress of technology today (consider for example DAOs) it makes more sense to have poolbeing.  In light of your question, "what is a poolbeing," I would ask to you in return, "What is a poolman?"  And there is nothing, because there is no poolman.  The poolman is not there.  We knock at the door, and he is not home.

'In Search of The Eternal Poolbeing'

"Because of the one-pointed Time awareness in which the conventional mind remains immersed, humans tend to think of everything in a sequential, word-oriented framework.  This mental trap produces very short-term concepts of effectiveness and consequences, a condition of constant, unplanned response to crises."

- Liet-Kynes
  The Arrakis Workbook
  (as quoted from Children of Dune, by Frank Herbert)

**Comment by sipa** [comment]
Not sure about the intention, but I always interpreted 'man' here as an
abbreviation for 'mamager'...

**Comment by dexX7** [comment]
The command `"invalidateblock"` is handy for testing with one node. I was looking for mempool manipulation, and this was the closest thing I've found, but in particular I was looking for a way to clear the whole mempool. So for what it's worth, since this is slightly related:
 Slightly related, so for what it's worth, I

**Comment by pstratem** [comment]
@ABISprotocol poolman is short for pool manager, manager is from the latin manus for hand, and has nothing to do with gender

**Comment by laanwj** [comment]
How to move forward here? (and I don't mean in regard to silly word games)

**Comment by pstratem** [comment]
Is expiration of mempool transactions on a fixed known schedule really the best solution?

**Comment by jgarzik** [comment]
@laanwj The merge blocker with this pull request is a @sipa comment from long ago:  The janitor must not sweep transactions which are "precious" to the local node.  Generally this means wallet transactions.

This is not a blocker for disabled-wallet configurations; thus two paths forward are,
1. Disable poolman for wallet nodes.
2. Do The Right Thing, and create an interface whereby the wallet module tells mempool (or poolman) which transactions are precious and should not be automatically culled.  (cc @jonasschnelli)

@pstratem No.  It is however a rough approximation of a solution that can be deployed today, modulo comments just made.

The consensus on "best" solution is a "superblock" type model, where
* Mempool is modelled after a X megabyte sized block (X = 24 hours worth of data, according to proposal)
* Transactions entering system are judged based on fee/priority, and insertion-sorted accordingly
* One or more transactions, perhaps include the latest transaction itself, if total mempool size exceeds X

This real-time is preferable to periodic garbage collection; however the "ideal solution" does not yet have any code or testing under its belt.

poolman is an imperfect stopgap solution until The Ideal Solution arrives.

**Comment by jonasschnelli** [comment]
For the wallet enabled mode, i think, if there would be a signal over CValidationInterface (something like `CheckJanitor(const CTransaction& tx, bool& clean)` the wallet could allow/disallow "precious" transactions to be preserved/not affected by the janitor.
I will rebase this PR and add the additional wallet/validation-interface signaling (during the next couple of days).
IMO this PR (or lets say something that clean the mempool) is long overdue.

**Comment by pstratem** [comment]
@jgarzik well clearly this a huge priority for merging (it's been sitting around for over a year now)

modeling the mempool as a superblock seems reasonable, but is maybe too expensive to actually run everytime a new transaction is included.

what do you think about simply removing the transaction (and it's dependents) with the lowest priority?

**Comment by ABISprotocol** [comment]
@pstratem removing transaction(s) and dependents with lowest priority is not a good idea because low-priority dust (and/or transactions which exceed dust threshold but which are of lower priorities) should not be rejected due to the importance of inclusion of types of transactions to support a wide range of participants' needs, income ranges, and re-envisioning of dust as microgiving capacity or opportunity ( see my comments in #6402 - #1536 - #6201 ).

**Comment by jgarzik** [comment]
Closing out older pull requests.

This is not a judgement on this pull request or time-based expiration.  Time-based mempool expiration continues to be the best way to sample what miners are not confirming.

There is potential for resubmitting or reopening this in the future.

**Pull request closed without merging**